File: SpeedTestGameClient/core/src/main/java/si/um/feri/speedii/dao/http/HttpEvent.java
package si.um.feri.speedii.dao.http;

import okhttp3.*;
import org.bson.types.ObjectId;
import org.json.JSONArray;
import org.json.JSONObject;

import io.github.cdimascio.dotenv.Dotenv;
import si.um.feri.speedii.classes.Event;
import si.um.feri.speedii.classes.Location;
import si.um.feri.speedii.classes.SessionManager;
import si.um.feri.speedii.dao.EventCRUD;

import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;


//TODO test this in case it is used
public class HttpEvent implements EventCRUD {

    private final String ip;
    private final OkHttpClient client;
    private final SessionManager sessionManager;

    public HttpEvent(SessionManager sessionManager) {
        this.sessionManager = sessionManager;

        // Load environment variables from .env file
        Dotenv dotenv = Dotenv.load();
        this.ip = dotenv.get("IP"); // Get the IP from the .env file

        this.client = new OkHttpClient();
    }

    @Override
    public Event getById(ObjectId id) {
        String url = ip + "/event/" + id.toString();
        Request request = new Request.Builder().url(url).get().build();

        try (Response response = client.newCall(request).execute()) {
            if (response.isSuccessful()) {
                String responseBody = response.body().string();
                JSONObject jsonObject = new JSONObject(responseBody);
                return parseEvent(jsonObject);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    @Override
    public List<Event> getAll() {
        String url = ip + "/event";
        Request request = new Request.Builder().url(url).build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Failed to execute request: " + response.code());
                return new ArrayList<>();
            }

            String responseBody = response.body().string();
            JSONArray jsonArray = new JSONArray(responseBody);
            List<Event> events = new ArrayList<>();

            for (int i = 0; i < jsonArray.length(); i++) {
                JSONObject jsonObject = jsonArray.getJSONObject(i);
                events.add(parseEvent(jsonObject));
            }

            return events;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return new ArrayList<>();
    }

    @Override
    public boolean insert(Event obj) {
        String url = ip + "/event";
        RequestBody requestBody = createRequestBody(obj);
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .post(requestBody)
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    @Override
    public boolean update(Event obj) {
        String url = ip + "/event/" + obj.getId().toString();
        RequestBody requestBody = createRequestBody(obj);
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .put(requestBody)
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    @Override
    public boolean delete(Event obj) {
        String url = ip + "/event/" + obj.getId().toString();
        Request request = new Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer " + sessionManager.getToken())
            .delete()
            .build();

        try (Response response = client.newCall(request).execute()) {
            return response.isSuccessful();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return false;
    }

    private RequestBody createRequestBody(Event obj) {
        JSONObject json = new JSONObject();
        json.put("name", obj.getName());
        json.put("type", obj.getType());
        json.put("time", obj.getTime().format(DateTimeFormatter.ISO_DATE_TIME));
        json.put("online", obj.isOnline());

        if (obj.getLocation() != null) {
            JSONObject locationObject = new JSONObject();
            locationObject.put("type", "Point");

            JSONArray coordinatesArray = new JSONArray();
            for (Double coordinate : obj.getLocation().coordinates) {
                coordinatesArray.put(coordinate);
            }

            locationObject.put("coordinates", coordinatesArray);
            json.put("location", locationObject);
        }

        return RequestBody.create(json.toString(), MediaType.parse("application/json"));
    }

    private Event parseEvent(JSONObject jsonObject) {
        String name = jsonObject.getString("name");
        String type = jsonObject.getString("type");

        String timeString = jsonObject.getString("time");
        DateTimeFormatter formatter = DateTimeFormatter.ISO_DATE_TIME;
        LocalDateTime time = LocalDateTime.parse(timeString, formatter);

        boolean online = jsonObject.getBoolean("online");

        Location location = null;
        if (!jsonObject.isNull("location") && jsonObject.get("location") instanceof JSONObject) {
            JSONObject locationObject = jsonObject.getJSONObject("location");

            List<Double> coordinates = new ArrayList<>();
            JSONArray coordinatesArray = locationObject.getJSONArray("coordinates");
            for (int i = 0; i < coordinatesArray.length(); i++) {
                coordinates.add(coordinatesArray.getDouble(i));
            }

            location = new Location(locationObject.getString("type"), coordinates);
        }

        ObjectId id = new ObjectId(jsonObject.getString("_id"));
        return new Event(name, type, time, online, location, id);
    }
}

File: MobileApp/app/speedtest/src/main/java/sosteric/pora/speedtest/dao/http/HttpEvent.kt
package dao.http

import Event
import Location
import SessionManager
import dao.EventCRUD
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.bson.types.ObjectId
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class HttpEvent(val sessionManager: SessionManager): EventCRUD{
    var ip = "http://20.160.43.196:3000"

    val client = OkHttpClient()
     override fun getById(id: ObjectId): Event? {
         val url = "${ip}/event/$id"
         val request = Request.Builder()
             .url(url)
             .get()
             .build()

         client.newCall(request).execute().use { response ->
             if (response.isSuccessful) {
                 val responseBody = response.body?.string()
                 responseBody?.let {
                     val jsonObject = JSONObject(it)
                     return parseEvent(jsonObject)
                 }
             }
             return null
         }
    }

    override fun getAll(): List<Event> {
        val client = OkHttpClient()
        val url = "${ip}/event"
        val request = Request.Builder()
            .url(url)
            .build()

        client.newCall(request).execute().use { response ->
            // Check if the request was successful (HTTP 200 OK)
            if (!response.isSuccessful) {
                // If not, handle the error
                println("Failed to execute request: ${response.code}")
                return emptyList()
            }

            // Get the response body as a string
            val responseBody = response.body?.string()

            responseBody?.let {
                val jsonArray = JSONArray(it)
                val events = mutableListOf<Event>()

                for (i in 0 until jsonArray.length()) {
                    val jsonObject = jsonArray.getJSONObject(i)

                    events.add(parseEvent(jsonObject))
                }

                return events
            }

            return emptyList()
        }
    }

    override fun insert(obj: Event): Boolean {
        val url = "$ip/event"
        val requestBody = createRequestBody(obj)
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .post(requestBody)
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }
    override fun update(obj: Event): Boolean {
        val url = "$ip/event/${obj.id}"
        val requestBody = createRequestBody(obj)
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .put(requestBody)
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }

    override fun delete(obj: Event): Boolean {
        val url = "$ip/event/${obj.id}"
        val request = Request.Builder()
            .url(url)
            .addHeader("authorization", "Bearer ${sessionManager.token}")
            .delete()
            .build()

        client.newCall(request).execute().use { response ->
            return response.isSuccessful
        }
    }

    private fun createRequestBody(obj: Event) : okhttp3.RequestBody {
        val json = JSONObject().apply {
            put("name", obj.name)
            put("type", obj.type)
            put("time", obj.time.format(DateTimeFormatter.ISO_DATE_TIME))
            put("online", obj.online)
            if(obj.location != null) {
                put("location", JSONObject().apply {
                    put("type", "Point")
                    put("coordinates", obj.location!!.coordinates)
                })
            }
        }
        return json.toString().toRequestBody("application/json".toMediaType())
    }

    private fun parseEvent(jsonObject: JSONObject) : Event{
        val name = jsonObject.getString("name")
        val type = jsonObject.getString("type")

        val timeString = jsonObject.getString("time")
        val formatter = DateTimeFormatter.ISO_DATE_TIME
        val time = LocalDateTime.parse(timeString, formatter)

        val online = jsonObject.getBoolean("online")

        val location= if (!jsonObject.isNull("location") && jsonObject.get("location") is JSONObject){
            val locationObject = jsonObject.getJSONObject("location")
            Location(
            type = locationObject.getString("type"),
            coordinates = locationObject.getJSONArray("coordinates")
                .let { coordinatesArray ->
                    List<Double>(coordinatesArray.length()) { coordinatesArray.getDouble(it) }
                }
            )
        } else null
        val id =ObjectId(jsonObject.getString("_id"))

        return Event(name, type, time, online, location, id)

    }

}